File: 01287. Element Appearing More Than 25% In Sorted Array/solution.js
/**
 * Problem: 1287. Element Appearing More Than 25% In Sorted Array
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Finds the integer appearing more than 25% in a sorted array
 *
 * @param {number[]} arr - Sorted input array
 *
 * @returns {number} - The special integer
 */
const findSpecialInteger = (arr) => {
  // Calculate the size of the input array
  const size = arr.length

  // Determine the threshold for 25% of the array length
  const qtr = Math.floor(size / 4)

  // Initialize a counter to track occurrences of the current number
  let cnt = 1

  // Initialize the previous number to the first element of the array
  let p = arr[0]

  // Iterate through the array starting from the second element
  for (let i = 1; i < size; i++) {
    // If the current number matches the previous, increment the counter
    if (p === arr[i]) cnt++
    // Otherwise, reset the counter to 1 for the new number
    else cnt = 1

    // If the counter exceeds the threshold, return the current number
    if (cnt > qtr) return arr[i]

    // Update the previous number to the current number
    p = arr[i]
  }

  // Return the last number checked (should not reach here for valid input)
  return p
}

File: 01287. Element Appearing More Than 25% In Sorted Array/solution.go
/**
 * Problem: 1287. Element Appearing More Than 25% In Sorted Array
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func findSpecialInteger(arr []int) int {
	// Initialize a counter to track the frequency of the current element
	cnt := 1

	// Iterate through the array starting from the second element
	for i := 1; i < len(arr); i++ {
		// Check if the current element is the same as the previous one
		if arr[i] == arr[i-1] {
			// Increment the counter if the elements are the same
			cnt++
		} else if cnt > len(arr)/4 {
			// If the counter exceeds 25% of the array length, return the element
			return arr[i-1]
		} else {
			// Reset the counter if the current element is different
			cnt = 1
		}
	}

	// Return the last element if it appears more than 25% of the time
	return arr[len(arr)-1]
}
